q2_demux/_demux.py

```diff
@@ -205,13 +209,26 @@ def emp(seqs: BarcodeSequenceFastqIterator,
                                                barcode_id=barcode_id,
                                                lane_number=1,
                                                read_number=1)
-            per_sample_fastqs[sample_id] = gzip.open(str(path), mode='w')
+            per_sample_fastqs[sample_id] = gzip.open(str(path), mode='a')
```

Comment:
Can you check whether the number of file descriptors is within the allowed threshold *before* opening a file here and in the conditional block below? It might make it a little clearer if the check happens first (i.e. check resource limit before requesting the resource), and it'll be more robust in case more than one file descriptor is opened in the loop (or if there are more open file descriptors than we expect, e.g. the manifest file). Since the threshold needs to be checked in a couple places, it's probably easiest to make a private helper that does this check and frees resources if necessary.

Thinking more about the threshold check, we don't know how many file descriptors are owned by the current process, so just checking against the soft limit using `resources.getrlimit` may not always work if there are other file descriptors open (e.g. the manifest opened above, or a file opened somewhere else in the process). It'll be more robust to check the current number of open file descriptors (across the entire process) against the soft limit. Using `psutil.num_fds` and comparing that to `resources.getrlimit` should get us what we want, and then `open_fh_count` won't be necessary. (`psutil` will become a dependency for the plugin, and it's worth adding a comment in the code that `psutil.num_fds` only works on Unix; we'll have to revisit when supporting Windows).